src/backend/gl/src/window/egl.rs

```diff
@@ -88,6 +99,70 @@ fn test_wayland_display() -> Option<libloading::Library> {
     Some(library)
 }
 
+#[cfg(not(target_os = "android"))]
+fn get_egl_display(
+    egl: &Starc<egl::DynamicInstance>,
+) -> Result<(egl::Display, Option<libloading::Library>), hal::UnsupportedBackend> {
+    // Query string with EGL_NO_DISPLAY is supported only if the EGL version is 1.5 or greater
+    let client_extensions = egl
+        .query_string(None, egl::EXTENSIONS)
+        .map_err(|_| hal::UnsupportedBackend)?
+        .to_string_lossy();
+    log::info!("Client extensions: {:?}", client_extensions);
+    let client_ext_list = client_extensions.split_whitespace().collect::<Vec<_>>();
+
+    let mut wsi_library = None;
+
+    let wayland_display = if client_ext_list.contains(&"EGL_EXT_platform_wayland") {
```

Comment:
could you provide the `client_ext_list` you get on Android? Does it even have those extensions? If it doesn't there is no need to gate this as `not(target_os = "android")`

Reply:
It would  be much better to have a run-time switch instead of the compile-time switch

Reply:
Are X11 or Wayland supports Android?

Query string with EGL_NO_DISPLAY is supported only if the EGL version is 1.5 or greater (Android 10+, < 10% devices).

Also we can get default display and make something like this:
```rust
let default_display = egl.get_display(egl::DEFAULT_DISPLAY).map_err(|_| hal::UnsupportedBackend)?;
let client_extensions = egl
    .query_string(Some(default_display), egl::EXTENSIONS)
    .map_err(|_| hal::UnsupportedBackend)?
    .to_string_lossy();
let client_ext_list = client_extensions.split_whitespace().collect::<Vec<_>>();

let wayland_display = if client_ext_list.contains(&"EGL_EXT_platform_wayland") {
    //...
}

let x11_display = if client_ext_list.contains(&"EGL_EXT_platform_x11") { 
    //...
}

display = if let Some(library) = wayland_display { 
    //...
}

display = if let Some(library) = x11_display { 
    //...
}
```

Reply:
I just want to minimize the code behind the feature gates, since it's more difficult to keep working.
For example, if `EGL_EXT_platform_wayland` and `EGL_EXT_platform_x11` are never exposed on Android anyway, there is no harm to have the code checking for them.

Reply:
Worked example without feature gates. But with some performance cost, because we should initialize default display and get client extension:

```rust
let default_display = egl.get_display(egl::DEFAULT_DISPLAY).ok_or_else(|| {
    log::warn!("Unable to get default display");
    hal::UnsupportedBackend
})?;

egl.initialize(default_display).map_err(|_| hal::UnsupportedBackend)?;

// Query string with EGL_NO_DISPLAY is supported only if the EGL version is 1.5 or greater
let client_extensions = egl
    .query_string(Some(default_display), egl::EXTENSIONS)
    .map_err(|_| hal::UnsupportedBackend)?
    .to_string_lossy();
log::info!("Client extensions: {:?}", client_extensions);
let client_ext_list = client_extensions.split_whitespace().collect::<Vec<_>>();

egl.terminate(default_display).map_err(|_| hal::UnsupportedBackend)?;

let mut wsi_library = None;

let wayland_display = if client_ext_list.contains(&"EGL_EXT_platform_wayland") {
    test_wayland_display()
} else {
    None
};

let x11_display = if client_ext_list.contains(&"EGL_EXT_platform_x11") {
    open_x_display()
} else {
    None
};
```

Reply:
These are different extensions. See:
> If display is EGL_NO_DISPLAY, then the EGL_EXTENSIONS string describes the set of supported client extensions. If display is a valid, initialized display, then the EGL_EXTENSIONS string describes the set of display extensions supported by that display. The set of supported client extensions is disjoint from the set of extensions supported by any given display. 

So the current code does it correctly. It doesn't however take into account that requirement for 1.5 EGL version. I don't know how we could do that, even, since we at this early point we don't know anything about the version. I think we should just check the error code and proceed without client extensions (i.e. empty list?..) if that fails without considering the backend unavailable.

Reply:
> These are different extensions

TIL. Fixed.